core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts

```diff
@@ -16,170 +16,169 @@
-         parsedResponse.error || 'Error communicating with server.');
-     });
-   }
-
-   clearEditor(): void {
-     this.topicSummaries = null;
-     this.rolesFetched = false;
-     this.selectingNewRole = false;
-     this.username = '';
-     this.userRoles = [];
-     this.possibleRolesToAdd = [];
-     this.managerInTopicsWithId = [];
-     this.roleInUpdate = null;
-     this.changingBannedVlaue = false;
-     this.userIsBanned = false;
-     this.startedEditingRole = false;
-   }
-
-   ngOnInit(): void {
-     this.roleToActions = null;
-     this.adminDataService.getDataAsync().then(adminDataObject => {
-       this.UPDATABLE_ROLES = adminDataObject.updatableRoles;
-       this.VIEWABLE_ROLES = adminDataObject.viewableRoles;
-       this.topicSummaries = adminDataObject.topicSummaries;
-       this.roleToActions = adminDataObject.roleToActions;
-     });
-   }
- }
-
- angular.module('oppia').directive('oppiaAdminRolesTab',
-   downgradeComponent({
-     component: AdminRolesTabComponent
-   }) as angular.IDirectiveFactory);
\ No newline at end of file
+import { Component, OnInit } from '@angular/core';
+import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
+import { downgradeComponent } from '@angular/upgrade/static';
+import { AdminDataService } from '../services/admin-data.service';
+import { AdminBackendApiService } from 'domain/admin/admin-backend-api.service';
+import { TopicManagerRoleEditorModalComponent } from './topic-manager-role-editor-modal.component';
+import { AlertsService } from 'services/alerts.service';
+
+@Component({
+  selector: 'oppia-admin-roles-tab',
+  templateUrl: './admin-roles-tab.component.html'
+})
+export class AdminRolesTabComponent implements OnInit {
+  UPDATABLE_ROLES = null;
+  VIEWABLE_ROLES = null;
+  topicSummaries = null;
+  roleToActions = null;
+  rolesFetched = false;
+
+  selectingNewRole = false;
+  username = '';
+  userRoles = [];
+  possibleRolesToAdd = [];
+  managerInTopicsWithId = [];
+  roleInUpdate = null;
+  errorMessage = null;
+  changingBannedVlaue = false;
+  userIsBanned = false;
+  startedEditingRole = false;
+
+  constructor(
+    private adminDataService: AdminDataService,
+    private adminBackendApiService: AdminBackendApiService,
+    private alertsService: AlertsService,
+    private modalService: NgbModal
+  ) {}
+
+  startEditing(): void {
+    this.startedEditingRole = true;
+    this.adminBackendApiService.viewUsersRoleAsync(
+      'username', null, this.username).then((userRoles) => {
+      this.rolesFetched = true;
+      this.userRoles = userRoles.roles;
+      this.managerInTopicsWithId = userRoles.managed_topic_ids;
+      this.userIsBanned = userRoles.banned;
+    });
+  }
+
+  showNewRoleSelector(): void {
+    this.possibleRolesToAdd = Object.keys(this.UPDATABLE_ROLES).filter(
+      role => !this.userRoles.includes(role)).sort();
+    this.selectingNewRole = true;
+  }
+
+  removeRole(roleToRemove: string): void {
+    this.roleInUpdate = roleToRemove;
```

Comment:
Why do we need to store this on the component? (I also don't really know what roleInUpdate means.)

Reply:
As we are going to show spinner around the role which we are adding/removing (i.e, in update).
roleInUpdate: The name of the role which is getting updated to user role i.e, either push or pop from users.roles field. Is it fine, I wasn't able to find a better name for this, do you have any suggestion?

Reply:
OK, maybe try this.roleCurrentlyBeingUpdatedInBackend. 

Also maybe add a comment wherever this is initialized explaining that this is for both additions and removals. Do we null it out once the update has completed?

Reply:
Done!
> Do we null it out once the update has completed?

Yes, we are doing it on line 90!